src/v/ssx/future-util.h

```diff
@@ -99,13 +99,88 @@ inline auto async_transform(Rng&& rng, Func&& func) {
     return async_transform(rng.begin(), rng.end(), std::forward<Func>(func));
 }
 
+/// \brief Run tasks synchronously in order and wait for completion only
+/// invoking futures one after a previous has completed (range version).
+/// Difference is that the results are expected to be a series of collections
+/// for which are then transformed into one single collection.
+///
+/// Given a range [\c begin, \c end) of objects, run \c func on each \c *i in
+/// the range, and return a \c future<> containing a \c std::vector<> of
+/// std::vector<> of values. These values are then unflattened into a single
+/// vector as the functions return value. In case any of the given tasks fails,
+/// one of the exceptions is returned by this function as a failed future.
+///
+/// \param begin an \c InputIterator designating the beginning of the range
+/// \param end an \c InputIterator designating the end of the range
+/// \param func Function to invoke with each element in the range (will be
+/// futurized if it doesn't return a \c future<> and expected to return a
+/// std::vector<> of values)
+/// \return a \c future<> containing a \c std::vector<>
+/// of the results of the function invocations that resolves when all the
+/// function invocations complete.  If one or more return an exception, the
+/// return value contains one of the exceptions.
+// clang-format off
+template<typename Iterator, typename Func>
+CONCEPT(requires requires(Func f, Iterator i) {
+    *i++;
+    { i != i } ->std::convertible_to<bool>;
+    seastar::futurize_invoke(f, *i).get0().begin();
+    seastar::futurize_invoke(f, *i).get0().end();
+})
+// clang-format on
+inline auto async_flat_transform(Iterator begin, Iterator end, Func&& func) {
+    using result_type = decltype(seastar::futurize_invoke(func, *begin).get0());
+    using value_type = typename result_type::value_type;
+    std::vector<value_type> res;
+    res.reserve(std::distance(begin, end));
+    return seastar::do_with(
+      std::move(res),
+      std::move(begin),
+      std::move(end),
+      [func{std::forward<Func>(func)}](
+        std::vector<value_type>& res, Iterator& begin, Iterator& end) mutable {
+          return seastar::do_for_each(
+                   begin,
+                   end,
+                   [&res, func{std::forward<Func>(func)}](auto&& val) mutable {
```

Comment:
i think here again i'm not certain about the forwarding reference and if its operating at intended

Reply:
If I swap this for a single reference, the project compiles with one exception, this snippet in async_adl.h, which only differs from other uses of `async_transform` in that we are iterating over a primitive data type. So after a bit of compiler hackery this was the only solution that compiled.

```
        return ss::do_with(
          boost::irange<size_t>(0, size), [&in](const auto& r) {
              return ssx::async_transform(r.begin(), r.end(), [&in](size_t) {
                  return async_adl<value_type>{}.from(in);
              });
          });
```

Reply:
i'm not sure that compiles vs doesn't-compile is the right argument here. playing around with a similar code snippet i don't see `val` ever being any type other than a l-value reference. can you provide a snippet of code that can be copy pasted that doesn't compile?

Reply:
note that in this case `func(*irange::begin())` passes an r-value ref to a temporary. other containers might return l-value ref, hence the need for `auto&&`.